Information/CenterOfMass.FCMacro

```diff
@@ -1,891 +1,966 @@
+        self.checkColorify.setIcon(QtGui.QIcon(':/icons/Std_RandomColor.svg'))
+        self.checkColorify.setIconSize(g_icon_size)
+        self.checkColorify.stateChanged.connect(self.on_stateChanged_Colorify)
 
-        massGroupBox = QtGui.QGroupBox()
         massLayout = QtGui.QHBoxLayout()
-        massLayout.addWidget(self.label_Mass)
-        massLayout.addWidget(self.resultMasseTot)
-        massLayout.addWidget(self.showCdg)
-        massLayout.addWidget(self.changeRadius)
-        massLayout.addWidget(self.colorify)
+        massLayout.addWidget(label_Mass)
+        massLayout.addWidget(self.resultMass)
+        massLayout.addWidget(self.checkColorify)
+        massGroupBox = QtGui.QGroupBox()
         massGroupBox.setLayout(massLayout)
-        massGroupBox.setFixedHeight(50)
         layout.addWidget(massGroupBox)
 
-        self.label_CdG = QtGui.QLabel(self)
-        self.label_CdG.setObjectName('label_CdG')
-        self.label_CdG.setText('Center of mass')
-        self.resultCdGX = QtGui.QLineEdit(self)
-        self.resultCdGX.setGeometry(QtCore.QRect(115, g_win_height - margin - 30, 65, 25))
-        self.resultCdGX.setObjectName('center of mass')
-        self.resultCdGX.setReadOnly(True)
-        self.resultCdGX.setPlaceholderText('no computation done')
-        self.resultCdGX.setToolTip('center of mass X (in %s)' % lengthU)
-
-        self.resultCdGY = QtGui.QLineEdit(self)
-        self.resultCdGY.setObjectName('center of mass')
-        self.resultCdGY.setReadOnly(True)
-        self.resultCdGY.setPlaceholderText('no computation done')
-        self.resultCdGY.setToolTip('center of mass Y (in %s)' % lengthU)
-
-        self.resultCdGZ = QtGui.QLineEdit(self)
-        self.resultCdGZ.setObjectName('center of mass')
-        self.resultCdGZ.setReadOnly(True)
-        self.resultCdGZ.setPlaceholderText('no computation done')
-        self.resultCdGZ.setToolTip('center of mass Z (in %s)' % lengthU)
-
-        cdgGroupBox = QtGui.QGroupBox()
-        cdgLayout = QtGui.QHBoxLayout()
-        cdgLayout.addWidget(self.label_CdG)
-        cdgLayout.addWidget(self.resultCdGX)
-        cdgLayout.addWidget(self.resultCdGY)
-        cdgLayout.addWidget(self.resultCdGZ)
-        cdgGroupBox.setLayout(cdgLayout)
-        cdgGroupBox.setFixedHeight(50)
-        layout.addWidget(cdgGroupBox)
+    def validSelection(self):
+        """Get valid objects (Shape, Mesh) from selection"""
+        def sortSelectionByLabel(s_):
+            return s_.Label
+        def sortSelectionByTree(s_):
+            return self.tree_list.index(s_.Label)
+
+        _sel = FreeCADGui.Selection.getSelection()
+
+        # Add contents of group objects 'groupObjs' (breadth-first search).
+        # For PartDesign objects a Group contains recursive features.
+        # Only nested App:Part are considered at this place.
+        groupObjs = ['App::DocumentObjectGroup', 'App::GeometryPython']
+        i_ = 0
+        while i_ < len(_sel):
```

Comment:
Can you please have a look at the depth-first search algorithm in `DeepCopy.FCMacro` and see if this is not possible to simplify the code here.

Reply:
I have several test cases where `DeepCopy.FCMacro` doesn't work as intended:
In the attached file neither for the inner nor the outer Part. There are too many copies.
[CDG_Test_File_schupin.zip](https://github.com/FreeCAD/FreeCAD-macros/files/8440529/CDG_Test_File_schupin.zip)